### PR TITLE
SE-0080 (4/4) - Failable initializers for Float->Int

### PR DESCRIPTION
**Pull request by ultramiraculous**
#### What's in this pull request?

Breaking out from #2742. Adds init?(exactly:) initializers for float types from int types.
#### Resolved bug number: ([SR-1491](https://bugs.swift.org/browse/SR-1491))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by ultramiraculous** [comment]
(Really late in the game here, I know)

**Comment by CodaFi** [comment]
Let's get a test run at least.

@swift-ci please smoke test.

**Comment by CodaFi** [comment]
// @shahmishal

**Comment by CodaFi** [comment]
@swift-ci please test.

File: stdlib/public/core/FloatingPointTypes.swift.gyb [hunk 1]
```diff
@@ -832,14 +816,29 @@ public prefix func - (x: ${Self}) -> ${Self} {
...
+%   ThatBuiltinName = self_ty.builtin_name
+%   srcBits = self_ty.bits
+%   sign = 's' if self_ty.is_signed else 'u'
   public init(_ v: ${That}) {
     _value = Builtin.${sign}itofp_${ThatBuiltinName}_FPIEEE${bits}(v._value)
   }
-% end
+  
+%   if srcBits < SignificandBitCount:
+  @available(*, message: "Converting ${That} to ${Self} will always succeed.")
```
Comment (CodaFi) [inline stdlib/public/core/FloatingPointTypes.swift.gyb]: Is there a test for this warning somewhere?

Comment (ultramiraculous) [inline stdlib/public/core/FloatingPointTypes.swift.gyb]: No... I hadn't thought to test these/I'm not sure I know how.

**Comment by jtbandes** [comment]
@ultramiraculous Any update on this? Looks like some conflicts have been introduced at this point. I'd be up for helping to resolve them if you like.

**Comment by ultramiraculous** [comment]
@jtbandes hey yeah sorry I at least completely dropped the ball on following up on this. Things got busy on my end. I can at least resolve the conflicts sometime today probably.

**Comment by ultramiraculous** [comment]
@gribozavr / @jtbandes Can this at least get the test suite run against it?

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by ultramiraculous** [comment]
Ping @CodaFi

File: stdlib/public/core/FloatingPointTypes.swift.gyb (hunk 1 above)
Comment (CodaFi) [inline stdlib/public/core/FloatingPointTypes.swift.gyb]: You can use GYB to write tests and the usual filecheck mechanism to check the warning.  [For example](https://github.com/apple/swift/blob/master/test/stdlib/Runtime.swift.gyb).

**Comment by CodaFi** [comment]
Okay, I can take responsibility for seeing this through.  I left a comment about testing the warning.  Because this addresses Dmitri's comments, it should be GTG after that.

**Comment by slavapestov** [comment]
@CodaFi Can we merge this?

**Comment by CodaFi** [comment]
This requires @ultramiraculous' attention still.

**Comment by CodaFi** [comment]
@ultramiraculous I'm going to rebase onto this and add the test myself.

**Comment by CodaFi** [comment]
@slavapestov Does this need to be PR'd over to 3.1 to?

**Comment by ultramiraculous** [comment]
Yeah sorry about that and thanks. I keep trying to find time to focus on this.

**Comment by CodaFi** [comment]
Well, on second thought I don't think the changes I want are enough to block this merge.

⛵️

**Pull request merged**